File: Domain/Device.cs
using System.ComponentModel.DataAnnotations;

namespace HostTool.Domain
{
    public class Device
    {
        [Key]
        public string DeviceId { get; set; } = string.Empty;
        
        [Required]
        [MaxLength(100)]
        public string DeviceName { get; set; } = string.Empty;
        
        [MaxLength(500)]
        public string? Description { get; set; }
        
        [Required]
        [MaxLength(50)]
        public string DeviceType { get; set; } = string.Empty; // IoT, Sensor, Controller, etc.
        
        [Required]
        [MaxLength(50)]
        public string Status { get; set; } = "Offline"; // Online, Offline, Error, Maintenance
        
        [MaxLength(100)]
        public string? Location { get; set; }
        
        [MaxLength(50)]
        public string? FirmwareVersion { get; set; }
        
        [MaxLength(50)]
        public string? HardwareVersion { get; set; }
        
        public string? LastKnownIP { get; set; }
        
        public DateTime? LastSeen { get; set; }
        
        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
        
        public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;
        
        // Device capabilities and configuration
        public string? Capabilities { get; set; } // JSON string for device capabilities
        public string? Configuration { get; set; } // JSON string for device configuration
        
        // Device health metrics
        public double? BatteryLevel { get; set; }
        public double? Temperature { get; set; }
        public double? SignalStrength { get; set; }
        
        // Device metadata
        public string? Manufacturer { get; set; }
        public string? Model { get; set; }
        public string? SerialNumber { get; set; }
        
        // Network information
        public string? MacAddress { get; set; }
        public int? Port { get; set; }
        
        // Device state
        public bool IsActive { get; set; } = true;
        public bool IsRegistered { get; set; } = false;
        
        // Heartbeat settings
        public int HeartbeatInterval { get; set; } = 30; // seconds
        public DateTime? LastHeartbeat { get; set; }
        
        // Error tracking
        public string? LastError { get; set; }
        public DateTime? LastErrorTime { get; set; }
        public int ErrorCount { get; set; } = 0;
    }
    
    public class DeviceEvent
    {
        [Key]
        public string EventId { get; set; } = string.Empty;
        
        [Required]
        public string DeviceId { get; set; } = string.Empty;
        
        [Required]
        [MaxLength(50)]
        public string EventType { get; set; } = string.Empty; // Register, Heartbeat, StatusChange, Error, etc.
        
        [MaxLength(500)]
        public string? Message { get; set; }
        
        public string? Data { get; set; } // JSON string for additional event data
        
        public DateTime Timestamp { get; set; } = DateTime.UtcNow;
        
        [MaxLength(20)]
        public string Severity { get; set; } = "Info"; // Info, Warning, Error, Critical
        
        public bool IsProcessed { get; set; } = false;
    }
    
    public class DeviceRegistrationRequest
    {
        public string DeviceName { get; set; } = string.Empty;
        public string? Description { get; set; }
        public string DeviceType { get; set; } = string.Empty;
        public string? Location { get; set; }
        public string? FirmwareVersion { get; set; }
        public string? HardwareVersion { get; set; }
        public string? Capabilities { get; set; }
        public string? Configuration { get; set; }
        public string? Manufacturer { get; set; }
        public string? Model { get; set; }
        public string? SerialNumber { get; set; }
        public string? MacAddress { get; set; }
        public int HeartbeatInterval { get; set; } = 30;
    }
    
    public class DeviceHeartbeatRequest
    {
        public string DeviceId { get; set; } = string.Empty;
        public string? Status { get; set; }
        public double? BatteryLevel { get; set; }
        public double? Temperature { get; set; }
        public double? SignalStrength { get; set; }
        public string? LastKnownIP { get; set; }
        public string? Data { get; set; } // Additional sensor data
    }
}

File: view/src/api/device.ts
import http from './http'

export interface Device {
  deviceId: string
  deviceName: string
  description?: string
  deviceType: string
  status: 'Online' | 'Offline' | 'Error' | 'Maintenance'
  location?: string
  firmwareVersion?: string
  hardwareVersion?: string
  lastKnownIP?: string
  lastSeen?: string
  createdAt: string
  updatedAt: string
  capabilities?: string
  configuration?: string
  batteryLevel?: number
  temperature?: number
  signalStrength?: number
  manufacturer?: string
  model?: string
  serialNumber?: string
  macAddress?: string
  port?: number
  isActive: boolean
  isRegistered: boolean
  heartbeatInterval: number
  lastHeartbeat?: string
  lastError?: string
  lastErrorTime?: string
  errorCount: number
}

export interface DeviceEvent {
  eventId: string
  deviceId: string
  eventType: string
  message?: string
  data?: string
  timestamp: string
  severity: 'Info' | 'Warning' | 'Error' | 'Critical'
  isProcessed: boolean
}

export interface DeviceRegistrationRequest {
  deviceName: string
  description?: string
  deviceType: string
  location?: string
  firmwareVersion?: string
  hardwareVersion?: string
  capabilities?: string
  configuration?: string
  manufacturer?: string
  model?: string
  serialNumber?: string
  macAddress?: string
  heartbeatInterval?: number
}

export interface DeviceHeartbeatRequest {
  deviceId: string
  status?: string
  batteryLevel?: number
  temperature?: number
  signalStrength?: number
  lastKnownIP?: string
  data?: string
}

export interface UpdateStatusRequest {
  status: string
}

export const deviceService = {
  // Register a new device
  register(request: DeviceRegistrationRequest) {
    return http.post<{ success: boolean; deviceId: string; message: string }>('/device/register', request)
  },

  // Send heartbeat
  heartbeat(request: DeviceHeartbeatRequest) {
    return http.post<{ success: boolean; message: string; timestamp: string }>('/device/heartbeat', request)
  },

  // Get all devices
  getAll() {
    return http.get<Device[]>('/device')
  },

  // Get device by ID
  getById(deviceId: string) {
    return http.get<Device>(`/device/${deviceId}`)
  },

  // Get online devices
  getOnline() {
    return http.get<Device[]>('/device/online')
  },

  // Get offline devices
  getOffline() {
    return http.get<Device[]>('/device/offline')
  },

  // Update device status
  updateStatus(deviceId: string, request: UpdateStatusRequest) {
    return http.put<{ success: boolean; message: string }>(`/device/${deviceId}/status`, request)
  },

  // Delete device
  delete(deviceId: string) {
    return http.delete<{ success: boolean; message: string }>(`/device/${deviceId}`)
  },

  // Get device events
  getEvents(deviceId: string, limit: number = 50) {
    return http.get<DeviceEvent[]>(`/device/${deviceId}/events?limit=${limit}`)
  },

  // Get device status
  getStatus(deviceId: string) {
    return http.get<{ deviceId: string; isOnline: boolean; timestamp: string }>(`/device/${deviceId}/status`)
  },

  // Cleanup offline devices
  cleanup(timeoutMinutes: number = 5) {
    return http.post<{ success: boolean; message: string }>(`/device/cleanup?timeoutMinutes=${timeoutMinutes}`)
  }
}
